base/channels.jl

```diff
@@ -253,6 +253,7 @@ Stacktrace:
 """
 function bind(c::Channel, task::Task)
     T = Task(() -> close_chnl_on_taskdone(task, c))
+    T.sticky = false
```

Comment:
this is okay, but the other change is invalid (it does not preserve the same behavior on both sides of the if statement)

Reply:
One side calls `schedule` explicitly, the other one implements an optimized copy of `schedule`

Reply:
The semantics of this `_wait2`, IIUC, is that `waiter` will run after `t` is done. The stickiness of `waiter` or `t` are not relevant here, so there is no need to check those or change anything. That's the part I removed.

Reply:
The semantics are that it is an optimized version of `schedule`, which defers the call until `t` is done, and `schedule` defines how stickiness works, so it is relevant here.

Reply:
[task_done_hook](https://github.com/JuliaLang/julia/blob/master/base/task.jl#L655) calls `notify(donenotify)`, which [calls](https://github.com/JuliaLang/julia/blob/master/base/condition.jl#L154) `schedule`, which, as you point out, addresses the sticky bit. It is redundant to check it here.

Reply:
The way it is now, `waiter` will be stuck to the thread running the task that called `_wait2`. With this change, `schedule` will stick `waiter` to `t`'s thread which seems more correct to me.

Reply:
Yes, that would result in it becoming sticky on the wrong thread, so we must call it immediately instead.

Reply:
I dropped that commit but did push the `unlock` of `donenotify` after the call to `jl_set_task_tid` to prevent a race on that.

Reply:
That function is implementing trylock, which is not race-y, and wouldn't be protected by a lock associated with an entirely different Task

Reply:
The race is on `jl_set_task_tid`. If, right when we unlock `donenotify`, it is notified and this `waiter` task is `schedule`d, `enq_work` will also [call `jl_set_task_tid`](https://github.com/JuliaLang/julia/blob/0cfd6f291876bbeb525a14a23fb996f224898496/base/task.jl#L780). Then it is possible for the task to have the wrong `tid` set in it.

Reply:
Oh, I see now. That seems a bit sloppy of us to push it into the work queue before we've fully initialized it though. I think we should move the `jl_set_task_tid` before we acquire the lock, so that it is sequenced-before that enq_work call even starts.

Reply:
I thought about that, but if `istaskdone(waiter)`, then we're calling `schedule` which will do the `jl_set_task_tid` correctly anyway.

Reply:
I still like the idea of having the lock-acquire as a lightweight memory barrier making that visible to other processes before we pointer-publish it. The cost should be cheap to repeat it, since schedule skips it if already done.

Reply:
Moved it.

Reply:
All okay now @vtjnash?